quantum/audio/audio_chibios_pwm.c

```diff
@@ -0,0 +1,195 @@
+/* Copyright 2020 Jack Humbert
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+Audio Driver: PWM
+
+the duty-cycle is always kept at 50%, and the pwm-period is adjusted to match the frequency of a note to be played back.
+
+this driver uses the chibios-PWM system to produce a squarewave on a given output pin.
+Either in software through a pwm callback and set/clear; or through the pwm hardware which directly toggles the pin via its alternate function. see your MCUs datasheet for pin can be driven by a timers T/IMx_CHy.
+
+ */
+
+
+/* STM32F103C Setup:
+halconf.h:
+#define HAL_USE_PWM                 TRUE
+#define HAL_USE_PAL                 TRUE
+#define HAL_USE_GPT                 TRUE
+
+mcuconf.h:
+#define STM32_PWM_USE_TIM1                  TRUE
+#define STM32_GPT_USE_TIM3                  TRUE
+
+used pin: PA8 (alternate0: Tim1_Ch1)
+
+from the datasheet for STM32F104C8B: alternate function of pin
+TIM1_CH1 = PA8
+TIM1_CH2 = PA9
+TIM1_CH3 = PA10
+TIM1_CH4 = PA11
+
+
+to use another hardware-pwm pin:
+- check the datasheet for the timer_channel and the alternate-function
+- replace PWMDx with the required timer
+- reorder the PCMConfig.channels to match the TIMx_CHy (the stm datasheets usually have CH1 to CH4)
+- set the pin to the appropiate alternate function (beware: differences in chibios GPIOv1, v2, ...)
+*/
+
+#include "audio.h"
+#include "ch.h"
+#include "hal.h"
+
+#include <string.h>
+#include "print.h"
+
+#if defined (AUDIO_DRIVER_PWM)
+#    if !defined(AUDIO_PIN)
```

Comment:
```suggestion
#    if !defined(AUDIO_PIN) && defined(A4_AUDIO)
#        define AUDIO_PIN A4
#    endif
#    if !defined(AUDIO_PIN) && defined(A5_AUDIO)
#        define AUDIO_PIN A5
#    endif
#    if !defined(AUDIO_PIN)
```

Reply:
Understood; the main rationale behind this request was purely for consistency with the other implementations. Trying to explain to people "oh, that style of `#define` really only applies to AVR, not ARM" is a bit more of an ongoing support issue that I'd like to avoid.

Reply:
since the whole audio-pin configuration didn't "feel quite right", 6239fbe184636bf7844b9a5ee3c36447080e9571 tries to resolve that mess with two makefile defines/configuration switches: AUDIO_PIN and AUDIO_PIN_ALT, which the user can/should set and then ripple out into the different audio-drivers as AUDIO_PIN_XY, with corresponding checks